src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,355 @@
+                }
+                request.Content = multipartContent;
+            }
+
+            return request;
+        }
+
+        public override bool IsReadRequest => false;
+
+        public void Dispose()
+        {
+            foreach (var command in _commands)
+                command?.Dispose();
+
+            if (_uniqueAttachmentStreams != null)
+            {
+                foreach (var uniqueAttachmentStream in _uniqueAttachmentStreams)
+                    uniqueAttachmentStream?.Dispose();
+            }
+
+            Result?.Dispose();
+
+            _returnCtx?.Dispose();
+        }
+    }
+
+    public class SingleShardedCommand
+    {
+        public string Id;
+        public Stream AttachmentStream;
+        public MemoryStream BufferedCommand;
+    }
+
+    public class ShardedBatchCommand : IEnumerator<SingleShardedCommand>, IEnumerable<SingleShardedCommand>
+    {
+        public List<MemoryStream> BufferedCommands;
+        public ArraySegment<BatchRequestParser.CommandData> ParsedCommands;
+        public List<Stream> AttachmentStreams;
+        public bool IsClusterTransaction;
+
+        private int _commandPosition;
+        private int _streamPosition;
+
+        public bool MoveNext()
+        {
+            if (_commandPosition == ParsedCommands.Count)
+                return false;
+
+            var cmd = ParsedCommands[_commandPosition];
+            var stream = cmd.Type == CommandType.AttachmentPUT ? AttachmentStreams[_streamPosition++] : null;
+
+            _current = new SingleShardedCommand
+            {
+                Id = cmd.Id,
+                AttachmentStream = stream,
+                BufferedCommand = BufferedCommands[_commandPosition]
+            };
+            _commandPosition++;
+            return true;
+        }
+
+        public void Reset()
+        {
+            throw new NotSupportedException();
+        }
+
+        private SingleShardedCommand _current;
+
+        public SingleShardedCommand Current => _current;
+
+        object IEnumerator.Current => _current;
+
+        public void Dispose()
+        {
+        }
+
+        public IEnumerator<SingleShardedCommand> GetEnumerator()
+        {
+            return this;
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+    }
+
+    
+
+    public class ShardedBatchCommandBuilder : BatchRequestParser.BatchCommandBuilder
+    {
+        public List<Stream> Streams;
+        public List<BufferedCommand> BufferedCommands = new List<BufferedCommand>();
+
+        public ShardedBatchCommandBuilder(RequestHandler handler, string database, char identityPartsSeparator) :
+            base(handler, database, identityPartsSeparator)
+        {
+        }
+
+        public override async Task SaveStream(JsonOperationContext context, Stream input)
+        {
+            Streams ??= new List<Stream>();
+
+            var ms = new MemoryStream();
```

Comment:
What if this is a big attachments? IIRC, we are writing these to disk , no?

Reply:
Yes, in the normal code path it is coupled with a specific database, so we use `_database.DocumentsStorage.AttachmentsStorage.GetTempFile` to store it.
Here we don't have it, so we should use the System temp path?

Reply:
That ends up here: 
`_documentsStorage.Environment.Options.DataPager.Options.TempPath.Combine(name);`

So we either have a `TempPath` in the database configuration or use the default one.
Note that we want one per sharded db, so we'll create a (mostly) empty directory for that.

Reply:
I think we might not have this info, since we need to the `StorageConfiguration` which we might not have.
So I put it in the server temp folder.